haystack/nodes/prompt/providers.py

```diff
@@ -486,3 +499,111 @@ def supports(cls, model_name_or_path: str, **kwargs) -> bool:
+            messages = [{"role": "user", "content": prompt}]
+        elif isinstance(prompt, list) and len(prompt) > 0 and isinstance(prompt[0], dict):
+            messages = prompt
+        else:
+            raise ValueError(
+                f"Provided prompt is incorrect."
+                f"Model {self.model_name_or_path} requires either a string or messages using ChatML format."
+                f"See https://github.com/openai/openai-python/blob/main/chatml.md for more details."
+            )
+
+        kwargs_with_defaults = self.model_input_kwargs
+        if kwargs:
+            # we use keyword stop_words but OpenAI uses stop
+            if "stop_words" in kwargs:
+                kwargs["stop"] = kwargs.pop("stop_words")
+            if "top_k" in kwargs:
+                top_k = kwargs.pop("top_k")
+                kwargs["n"] = top_k
+                kwargs["best_of"] = top_k
+            kwargs_with_defaults.update(kwargs)
+        payload = {
+            "model": self.model_name_or_path,
+            "messages": messages,
+            "max_tokens": kwargs_with_defaults.get("max_tokens", self.max_length),
+            "temperature": kwargs_with_defaults.get("temperature", 0.7),
+            "top_p": kwargs_with_defaults.get("top_p", 1),
+            "n": kwargs_with_defaults.get("n", 1),
+            "stream": False,  # no support for streaming
+            "stop": kwargs_with_defaults.get("stop", None),
+            "presence_penalty": kwargs_with_defaults.get("presence_penalty", 0),
+            "frequency_penalty": kwargs_with_defaults.get("frequency_penalty", 0),
+            "logit_bias": kwargs_with_defaults.get("logit_bias", {}),
+        }
+        response = openai_request(url=self.url, headers=self.headers, payload=payload)
+        _check_openai_finish_reason(result=response, payload=payload)
+        assistant_response = [choice["message"]["content"].strip() for choice in response["choices"]]
+
+        if "stop" in kwargs_with_defaults and kwargs_with_defaults["stop"] is not None:
+            # Although ChatGPT generates text until stop words are encountered, unfortunately it includes the stop word
+            # We want to exclude it to be consistent with other invocation layers
+            stop_words = kwargs_with_defaults["stop"]
+            for idx, _ in enumerate(assistant_response):
+                for stop_word in stop_words:
+                    assistant_response[idx] = assistant_response[idx].replace(stop_word, "").strip()
+
+        return assistant_response
+
+    def _ensure_token_limit(self, prompt: Union[str, List[Dict[str, str]]]) -> Union[str, List[Dict[str, str]]]:
+        """Ensure that the length of the prompt and answer is within the max tokens limit of the model.
+        If needed, truncate the prompt text so that it fits within the limit.
+
+        :param prompt: Prompt text to be sent to the generative model.
+        """
+        if isinstance(prompt, str):
```

Comment:
Yes, it'll happen for sure...Maybe I missed something. Please let me know.

Reply:
Same as above...

Reply:
This is good. We can have both string prompts, and message dict prompts, but only in ChatGPTInvocationLayer; not in other implementations. This is good stuff now.